### PR TITLE
Thermostat fixes+updates 1

### LINKED ISSUE #1163: single point thermostats should work after power up without needing to be changed into a new mode
**Describe the problem you have/What new integration you would like**

see https://esphome.io/components/climate/thermostat.html
Single point thermostats (those with either **default_target_temperature_low** or **default_target_temperature_high** set but not both) should default to heat operation or cool operation so that they'll function correctly after a power loss, or they should function correctly under "heat/cool" operation.

**Please describe your use case for this integration and alternatives you've tried:**

When single point (cooling only or heating only) thermostats turn on they default to "heat/cool" operation and in that mode they won't activate.  They do work correctly after being switched to "heat" or "cool" operation.  I was able to work around this by setting both a **default_target_temperature_low** and **default_target_temperature_high** even though I only want cooling operation - I made the heating temperature arbitrarily low, and now in the default "heat/cool" mode it does work.  I assume using bang-bang would be an option as well.

**Additional context**

### LINKED ISSUE #1693: Climate hvac_action does not return to idle after being off
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Hass.io (HassOs Homeassistant)
**ESP (ESP32/ESP8266, Board/Sonoff):**

Wemos D1 mini ESP8266
**ESPHome version (latest production, beta, dev branch)**

Latest production
**Affected component:**

https://esphome.io/components/climate/thermostat.html

**Description of problem:**
When I open my windows, I turn off my esphome climate devices. When the window closes, I call the climate.set_hvac_mode to heat (via the lovelace interface)

When the current temperature is within the target temperature range, the hvac_action of the climate device does not change to idle, but remains off.

This causes some inconsistencies in the frontend and complicates the posibilities to trigger notifications based on hvac_action change
climate_hvac_action
**Problem-relevant YAML-configuration entries:**
```yaml
climate:
  - platform: thermostat
    name: Woonkamer
    sensor: temperatuur_woonkamer
    default_target_temperature_low: 21 °C
    heat_action:
      - switch.turn_on: woonkamer
    idle_action:
      - switch.turn_off: woonkamer
    hysteresis: 0.2
    visual:
      min_temperature: 15
      max_temperature: 25

switch:
  - platform: gpio
    pin: 
      number: 5
      inverted: true
    name: "Verwarming Woonkamer"
    id: woonkamer

sensor:
  - platform: homeassistant
    entity_id: sensor.temperature_158d00022c8b84
    id: temperatuur_woonkamer
    unit_of_measurement: "°C"
    accuracy_decimals: 1
```

**Logs (if applicable):**

```
[11:48:01][C][thermostat.climate:509]: Thermostat 'Woonkamer'
[11:48:01][C][thermostat.climate:514]:   Default Target Temperature Low: 21.0°C
[11:48:01][C][thermostat.climate:522]:   Hysteresis: 0.2°C
[11:48:01][C][thermostat.climate:523]:   Supports AUTO: NO
[11:48:01][C][thermostat.climate:524]:   Supports COOL: NO
[11:48:01][C][thermostat.climate:525]:   Supports DRY: NO
[11:48:01][C][thermostat.climate:526]:   Supports FAN_ONLY: NO
[11:48:01][C][thermostat.climate:527]:   Supports HEAT: YES
[11:48:01][C][thermostat.climate:528]:   Supports FAN MODE ON: NO
[11:48:01][C][thermostat.climate:529]:   Supports FAN MODE OFF: NO
[11:48:01][C][thermostat.climate:530]:   Supports FAN MODE AUTO: NO
[11:48:01][C][thermostat.climate:531]:   Supports FAN MODE LOW: NO
[11:48:01][C][thermostat.climate:532]:   Supports FAN MODE MEDIUM: NO
[11:48:01][C][thermostat.climate:533]:   Supports FAN MODE HIGH: NO
[11:48:01][C][thermostat.climate:534]:   Supports FAN MODE MIDDLE: NO
[11:48:01][C][thermostat.climate:535]:   Supports FAN MODE FOCUS: NO
[11:48:01][C][thermostat.climate:536]:   Supports FAN MODE DIFFUSE: NO
[11:48:01][C][thermostat.climate:537]:   Supports SWING MODE BOTH: NO
[11:48:01][C][thermostat.climate:538]:   Supports SWING MODE OFF: NO
[11:48:01][C][thermostat.climate:539]:   Supports SWING MODE HORIZONTAL: NO
[11:48:01][C][thermostat.climate:540]:   Supports SWING MODE VERTICAL: NO
[11:48:01][C][thermostat.climate:541]:   Supports TWO SET POINTS: NO
[11:48:01][C][thermostat.climate:542]:   Supports AWAY mode: NO

[11:50:26][D][climate:010]: 'Woonkamer' - Setting
[11:50:26][D][climate:014]:   Mode: OFF
[11:50:26][D][climate:262]: 'Woonkamer' - Sending state:
[11:50:26][D][climate:265]:   Mode: OFF
[11:50:26][D][climate:267]:   Action: OFF
[11:50:26][D][climate:276]:   Current Temperature: 22.40°C
[11:50:26][D][climate:282]:   Target Temperature: 22.00°C
[11:50:28][D][climate:010]: 'Woonkamer' - Setting
[11:50:28][D][climate:014]:   Mode: HEAT
[11:50:28][D][climate:262]: 'Woonkamer' - Sending state:
[11:50:28][D][climate:265]:   Mode: HEAT
[11:50:28][D][climate:267]:   Action: OFF
[11:50:28][D][climate:276]:   Current Temperature: 22.40°C
[11:50:28][D][climate:282]:   Target Temperature: 22.00°C

```

**Additional information and things you've tried:**
If I increase the target temperature, and then lower the target temperature. This is the log. In this case the Action does return to Idle
[11:52:12][D][climate:267]:   Action: IDLE
```
[11:52:09][D][climate:262]: 'Woonkamer' - Sending state:
[11:52:09][D][climate:265]:   Mode: HEAT
[11:52:09][D][climate:267]:   Action: HEATING
[11:52:09][D][climate:276]:   Current Temperature: 22.40°C
[11:52:09][D][climate:282]:   Target Temperature: 22.20°C
[11:52:12][D][climate:010]: 'Woonkamer' - Setting
[11:52:12][D][climate:025]:   Target Temperature: 22.00
[11:52:12][D][switch:025]: 'Verwarming Woonkamer' Turning OFF.
[11:52:12][D][switch:045]: 'Verwarming Woonkamer': Sending state OFF
[11:52:12][D][climate:262]: 'Woonkamer' - Sending state:
[11:52:12][D][climate:265]:   Mode: HEAT
[11:52:12][D][climate:267]:   Action: IDLE
[11:52:12][D][climate:276]:   Current Temperature: 22.40°C
[11:52:12][D][climate:282]:   Target Temperature: 22.00°C
```

### LINKED ISSUE #1758: Thermostat Away Setting Overrides Other Settings in climate.control
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Installed using pip

**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32 Wemos D1 Mini32

**ESPHome version (latest production, beta, dev branch)**

1.15.3

**Affected component:**

Thermostat

**Description of problem:**
When using climate.control to set the away configuration of the thermostat, any other temperature configuration settings in the command are ignored. For example, if I set the away target temperature as well as the away setting, then the target_temperature is set to the default value as defined in the thermostat component config, and not the value provided in climate.control.

I believe that the target temperature should be set to that defined in the control action, regardless of whether the away setting is set or not. I suspect that it also happens for the min and max temperature settings, but I have not tested this. I haven't come across documentation that describes the expected outcome for this set of climate.control parameters, so apologies in advanced if I've missed it.

In the code below, I've set up a sequence of events to show the effect (to clarify, this isn't how it's written in my application, but I thought it easier to provide simple code to demonstrate it). I've left out the usual wifi/ota/logger config for brevity - there is nothing unusual there.

Note that setting the away value and the target temperature separately works, and is a potential workaround. However, the setting of the away value alone could instigate an heat/idle action depending on the default temperature value, and therefore could lead to undesirable effect of quickly switching between the default and the correct target temperatures. For this reason, it would be good to fix this issue.

**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
  name: heatingcontrol
  platform: ESP32
  board: wemos_d1_mini32

  on_boot:
    - delay: 10s
    # This should set the target temperature to 12.3 (it does)
    - climate.control:
        id: zone_1_climate_control
        target_temperature: 12.3
    - delay: 10s
    # This should set the thermostat to 'away', using the default target temperature (it does)
    - climate.control:
        id: zone_1_climate_control
        away: true
    - delay: 10s
    # This should set the thermostat to 'home' (it does) and set the target temperature to 23.4 (it doesn't)
    - climate.control:
        id: zone_1_climate_control
        away: false
        target_temperature: 23.4
    - delay: 10s
    # This should set the thermostat to 'away' (it does) and set the target temperature to 8.9 (it doesn't)
    - climate.control:
        id: zone_1_climate_control
        away: true
        target_temperature: 8.9
    - delay: 10s
    # This sequence should set the thermostat to 'home' and set the target temperature to 23.4 (it does both)
    - climate.control:
        id: zone_1_climate_control
        away: false
    - climate.control:
        id: zone_1_climate_control
        target_temperature: 23.4

switch:
  - platform: gpio
    id: zone_1_radiator
    internal: true
    restore_mode: ALWAYS_OFF
    pin:
      number: D1

sensor:
  - platform: template
    id: test_temp_sensor
    internal: true
    lambda: |-
      return 18.0;

climate:
  - platform: thermostat
    id: zone_1_climate_control
    name: Zone 1 Heating Controller
    sensor: test_temp_sensor
    default_target_temperature_low: 10.0
    visual:
      min_temperature: 5.0
      max_temperature: 30.0
      temperature_step: 0.1
    away_config:
        default_target_temperature_low: 7.0
    heat_action:
      switch.turn_on: zone_1_radiator
    idle_action:
      switch.turn_off: zone_1_radiator

```

**Logs (if applicable):**

```
[17:31:32][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:31:32][D][climate:025]:   Target Temperature: 12.30
[17:31:32][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:31:32][D][climate:265]:   Mode: OFF
[17:31:32][D][climate:267]:   Action: OFF
[17:31:32][D][climate:276]:   Current Temperature: 18.00°C
[17:31:32][D][climate:282]:   Target Temperature: 12.30°C
[17:31:32][D][climate:285]:   Away: ON
[17:31:42][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:31:42][D][climate:034]:   Away Mode: ON
[17:31:42][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:31:42][D][climate:265]:   Mode: OFF
[17:31:42][D][climate:267]:   Action: OFF
[17:31:42][D][climate:276]:   Current Temperature: 18.00°C
[17:31:42][D][climate:282]:   Target Temperature: 7.00°C
[17:31:42][D][climate:285]:   Away: ON
[17:31:52][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:31:52][D][climate:025]:   Target Temperature: 23.40
[17:31:52][D][climate:034]:   Away Mode: OFF
[17:31:52][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:31:52][D][climate:265]:   Mode: OFF
[17:31:52][D][climate:267]:   Action: OFF
[17:31:52][D][climate:276]:   Current Temperature: 18.00°C
[17:31:52][D][climate:282]:   Target Temperature: 10.00°C
[17:31:52][D][climate:285]:   Away: OFF
[17:31:58][D][sensor:092]: 'test_temp_sensor': Sending state 18.00000  with 1 decimals of accuracy
[17:31:58][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:31:58][D][climate:265]:   Mode: OFF
[17:31:58][D][climate:267]:   Action: OFF
[17:31:58][D][climate:276]:   Current Temperature: 18.00°C
[17:31:58][D][climate:282]:   Target Temperature: 10.00°C
[17:31:58][D][climate:285]:   Away: OFF
[17:32:02][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:32:02][D][climate:025]:   Target Temperature: 8.90
[17:32:02][D][climate:034]:   Away Mode: ON
[17:32:02][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:32:02][D][climate:265]:   Mode: OFF
[17:32:02][D][climate:267]:   Action: OFF
[17:32:02][D][climate:276]:   Current Temperature: 18.00°C
[17:32:02][D][climate:282]:   Target Temperature: 7.00°C
[17:32:02][D][climate:285]:   Away: ON
[17:32:13][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:32:13][D][climate:034]:   Away Mode: OFF
[17:32:13][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:32:13][D][climate:265]:   Mode: OFF
[17:32:13][D][climate:267]:   Action: OFF
[17:32:13][D][climate:276]:   Current Temperature: 18.00°C
[17:32:13][D][climate:282]:   Target Temperature: 10.00°C
[17:32:13][D][climate:285]:   Away: OFF
[17:32:13][D][climate:010]: 'heatingcontrol Zone 1 Heating Controller' - Setting
[17:32:13][D][climate:025]:   Target Temperature: 23.40
[17:32:13][D][climate:262]: 'heatingcontrol Zone 1 Heating Controller' - Sending state:
[17:32:13][D][climate:265]:   Mode: OFF
[17:32:13][D][climate:267]:   Action: OFF
[17:32:13][D][climate:276]:   Current Temperature: 18.00°C
[17:32:13][D][climate:282]:   Target Temperature: 23.40°C
[17:32:13][D][climate:285]:   Away: OFF
```

**Additional information and things you've tried:**
Looking at the source code components/thermostat/thermostat_climate.cpp, I think the cause is here: https://github.com/esphome/esphome/blob/fbc1b3e316590776dc87dc90e1a98bb781517fde/esphome/components/thermostat/thermostat_climate.cpp#L55.
If that if-block was at the beginning of the ThermostatClimate::control method, then any additional parameters provided by the calling function would override the default values given by the ThermostatClimate::change_away_ function. I have not tested this, however, so it is only a suggestion for a fix.

Thank you for looking.

### LINKED ISSUE #2233: Thermostat doesn't start with a single temperature.
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

N/A

**ESP (ESP32/ESP8266, Board/Sonoff):**

N/A

**ESPHome version (latest production, beta, dev branch)**

1.19

**Affected component:**

Thermostat

**Description of problem:**
If you create a thermostat component with only one temperature, it will always be in the OFF action unless you switch the mode remotely.  This is because setup sets the mode to HEATCOOL (formerly AUTO) and that's not a valid condition in the compute action function, so the action nevers gets changed from OFF.

**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE

```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE

```

**Additional information and things you've tried:**

A workaround is to set the mode to HEAT or COOL (depending on which temperature is configured) in on_boot.

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix? 

Fixes issues listed below and adds a new action triggered by target temperature changes.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):**
Fixes https://github.com/esphome/issues/issues/1693
Fixes https://github.com/esphome/issues/issues/1758
Fixes https://github.com/esphome/issues/issues/2233
Fixes https://github.com/esphome/feature-requests/issues/1163

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/1311

## Test Environment

- [x] ESP32
- [ ] ESP8266

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/thermostat/climate.py [hunk 1]
```diff
@@ -141,6 +143,46 @@ def validate_thermostat(config):
...
+            "{} is set to {} but {} is not present in the configuration".format(
+                CONF_DEFAULT_MODE, CONF_FAN_ONLY_MODE, CONF_FAN_ONLY_ACTION
+            )
+        )
+    if config[CONF_DEFAULT_MODE] == CONF_HEAT_MODE and CONF_HEAT_ACTION not in config:
+        raise cv.Invalid(
+            "{} is set to {} but {} is not present in the configuration".format(
+                CONF_DEFAULT_MODE, CONF_HEAT_MODE, CONF_HEAT_ACTION
+            )
+        )
```
Comment (OttoWinter) [inline esphome/components/thermostat/climate.py]:
```suggestion
    # verify default climate mode is valid given above configuration
    default_mode = config[CONF_DEFAULT_MODE]
    requirements = {
        CONF_AUTO_MODE: [CONF_COOL_ACTION, CONF_HEAT_ACTION],
        CONF_COOL_MODE: [CONF_COOL_ACTION],
        CONF_HEAT_MODE: [CONF_HEAT_ACTION],
        CONF_DRY_MODE: [CONF_DRY_ACTION],
        CONF_FAN_ONLY_MODE: [CONF_FAN_ONLY_ACTION],
    }.get(default_mode, [])
    for req in requirements:
        if req not in config:
            raise cv.Invalid(
                f"{CONF_DEFAULT_MODE} is set to {default_mode} but {req} is not present in the configuration"
            )
```

Makes the code easier to read :)

File: esphome/components/thermostat/climate.py [hunk 2]
```diff
@@ -204,6 +246,17 @@ def validate_thermostat(config):
...
+                CONF_TARGET_TEMPERATURE_CHANGE_ACTION
+            ): automation.validate_automation(single=True),
+            cv.Optional(CONF_DEFAULT_MODE, default=CONF_OFF_MODE): cv.one_of(
+                CONF_AUTO_MODE,
+                CONF_COOL_MODE,
+                CONF_DRY_MODE,
+                CONF_FAN_ONLY_MODE,
+                CONF_HEAT_MODE,
+                CONF_OFF_MODE,
+            ),
```
Comment (jesserockz) [inline esphome/components/thermostat/climate.py]: You don't apply the CONF_DEFAULT_MODE value in codegen...

Comment (jesserockz) [reply esphome/components/thermostat/climate.py]: Also, best to set it up as a dict higher up with the actual enums as values and use `cv.enum` so you get the c++ enum value directly in `config[CONF_DEFAULT_MODE]`

File: esphome/components/thermostat/climate.py (hunk 1 above)
Comment (kbx81) [reply esphome/components/thermostat/climate.py]: Awesome! Thank you @OttoWinter...I'm a Python n00b so tips and tricks like this are very much appreciated! 😄

File: esphome/components/thermostat/climate.py (hunk 2 above)
Comment (kbx81) [reply esphome/components/thermostat/climate.py]: Hahaha...this is why we do PRs...and why I shouldn't stay up all night writing code on Friday and Saturday nights. Thanks again @jesserockz ...I'll fix this up. 😄

**Review by jesserockz — APPROVED** [review]

**Pull request merged**